File: packages/cli-js/src/utils/url-fetch.ts
/**
 * Extract http(s) URLs from text. Notion URLs (notion.so, notion.site) are excluded
 * so the CLI does not fetch them (use Cursor /p with MCP for Notion).
 */
const URL_RE = /https?:\/\/[^\s]+/g;
const NOTION_HOST_RE = /^https?:\/\/([^/]*\.)?(notion\.(so|site))(\/|$)/i;

function trimTrailingPunctuation(url: string): string {
  return url.replace(/[.,;:)!?\]]+$/, "");
}

export function extractUrlsFromGoal(goal: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  let m: RegExpExecArray | null;
  URL_RE.lastIndex = 0;
  while ((m = URL_RE.exec(goal)) !== null) {
    const raw = m[0];
    const url = trimTrailingPunctuation(raw);
    if (NOTION_HOST_RE.test(url)) continue;
    if (!seen.has(url)) {
      seen.add(url);
      out.push(url);
    }
  }
  return out;
}

const FETCH_TIMEOUT_MS = 10_000;

/**
 * Fetch URL and return body as plain text / simplified markdown. Returns empty string on failure.
 */
export async function fetchUrlContent(url: string, timeoutMs: number = FETCH_TIMEOUT_MS): Promise<string> {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: { "User-Agent": "PlanForge-CLI/1.0" },
    });
    if (!res.ok) return "";
    const text = await res.text();
    return text.slice(0, 100_000);
  } catch {
    return "";
  } finally {
    clearTimeout(t);
  }
}

/**
 * Fetch all non-Notion URLs from goal and return a single markdown block for context, or undefined if none.
 */
export async function fetchUrlsContext(goal: string): Promise<string | undefined> {
  const urls = extractUrlsFromGoal(goal);
  if (urls.length === 0) return undefined;
  const blocks: string[] = [];
  for (const url of urls) {
    const content = await fetchUrlContent(url);
    if (!content.trim()) continue;
    blocks.push(`## ${url}\n\n${content.trim().slice(0, 50_000)}`);
  }
  if (blocks.length === 0) return undefined;
  return "---\n\n### Fetched URLs\n\n" + blocks.join("\n\n");
}

File: packages/cli-py/planforge/utils/url_fetch.py
"""Extract http(s) URLs from goal and fetch content for plan context. Notion URLs are skipped (use Cursor /p with MCP)."""

import re
import urllib.error
import urllib.request

URL_RE = re.compile(r"https?://[^\s]+")
NOTION_HOST_RE = re.compile(r"^https?://([^/]*\.)?(notion\.(?:so|site))(\/|$)", re.IGNORECASE)
FETCH_TIMEOUT_S = 10
MAX_BODY = 100_000


def _trim_trailing_punctuation(url: str) -> str:
    return re.sub(r"[.,;:)!?\]]+$", "", url)


def extract_urls_from_goal(goal: str) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for m in URL_RE.finditer(goal):
        raw = m.group(0)
        url = _trim_trailing_punctuation(raw)
        if NOTION_HOST_RE.search(url):
            continue
        if url not in seen:
            seen.add(url)
            out.append(url)
    return out


def _fetch_url_content(url: str, timeout: int = FETCH_TIMEOUT_S) -> str:
    req = urllib.request.Request(url, headers={"User-Agent": "PlanForge-CLI/1.0"})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            body = resp.read().decode("utf-8", errors="replace")
            return body[:MAX_BODY]
    except (urllib.error.URLError, OSError, ValueError):
        return ""


def fetch_urls_context(goal: str) -> str | None:
    """Fetch all non-Notion URLs from goal and return a single markdown block for context, or None if none."""
    urls = extract_urls_from_goal(goal)
    if not urls:
        return None
    blocks: list[str] = []
    for url in urls:
        content = _fetch_url_content(url).strip()
        if not content:
            continue
        blocks.append(f"## {url}\n\n{content[:50_000]}")
    if not blocks:
        return None
    return "---\n\n### Fetched URLs\n\n" + "\n\n".join(blocks)
